src/probability/martingale.lean

```diff
@@ -353,6 +386,73 @@ end
 
 end submartingale
 
+-- We now prove the converse of the optional stopping theorem
+
+section piecewise_const
+
+variables {i j n : ℕ} {s : set α} {x : α} [decidable_pred (∈ s)]
+
+lemma piecewise_const_le_max : s.piecewise (λ _, i) (λ _, j) x ≤ max i j :=
+by { by_cases hx : x ∈ s; simp [hx], }
+
+@[measurability]
+lemma measurable_piecewise_const {m : measurable_space α} (hs : measurable_set s) :
+  measurable (s.piecewise (λ _, i) (λ _, j)) :=
+measurable.piecewise hs measurable_const measurable_const
+
+lemma is_stopping_time_piecewise_const (hij : i ≤ j) (hs : measurable_set[𝒢 i] s) :
+  is_stopping_time 𝒢 (s.piecewise (λ _, i) (λ _, j)) :=
+begin
+  refine is_stopping_time_of_measurable_set_eq (λ n, _),
+  by_cases hij' : i = j,
+  { simp [hij'], },
+  by_cases hin : i = n,
+  { have hs_n : measurable_set[𝒢 n] s, by { convert hs, exact hin.symm, },
+    exact (measurable_piecewise_const hs_n) (measurable_set_singleton n), },
+  by_cases hjn : j = n,
+  { have hs_n : measurable_set[𝒢 n] s, by { convert 𝒢.mono hij _ hs, exact hjn.symm, },
+    exact measurable_piecewise_const hs_n (measurable_set_singleton n), },
+  simp [set.piecewise, ite_eq_iff, hin, hjn],
+end
```

Comment:
Possibly slightly nicer as:
```suggestion
lemma is_stopping_time_piecewise_const (hij : i ≤ j) (hs : measurable_set[𝒢 i] s) :
  is_stopping_time 𝒢 (s.piecewise (λ _, i) (λ _, j)) :=
begin
  have hs_j : measurable_set[𝒢 j] s := 𝒢.mono hij _ hs,
  refine is_stopping_time_of_measurable_set_eq (λ n, _),
  obtain rfl | hij' := decidable.eq_or_ne i j,
  { simp, },
  obtain rfl | hin := decidable.eq_or_ne i n,
  { refine measurable.piecewise hs _ _ (measurable_set_singleton i),
    apply measurable_const,
    apply measurable_const, },
  obtain rfl | hjn := decidable.eq_or_ne j n,
  { refine measurable.piecewise hs_j _ _ (measurable_set_singleton j),
    apply measurable_const,
    apply measurable_const, },
  simp [set.piecewise, ite_eq_iff, hin, hjn],
end
```
since I'm not sure how useful `measurable_piecewise_const` is as a lemma.

Perhaps a more useful lemma would be a version of `measurable_const` that doesn't use typeclass search to find its measurable_space, which would eliminate those `apply`s.